exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Queries.kt

```diff
@@ -489,35 +484,41 @@ fun <T : Table> T.updateReturning(
  *
  * @param keys (optional) Columns to include in the condition that determines a unique constraint match.
  * If no columns are provided, primary keys will be used. If the table does not have any primary keys, the first unique index will be attempted.
+ * @param onUpdate Lambda block with an [UpdateStatement] as its argument, allowing values to be assigned to the UPDATE clause.
+ * To specify manually that the insert value should be used when updating a column, for example within an expression
+ * or function, invoke `insertValue()` with the desired column as the function argument.
+ * If left null, all columns will be updated with the values provided for the insert.
  * @param onUpdateExclude List of specific columns to exclude from updating.
  * If left null, all columns will be updated with the values provided for the insert.
  * @param where Condition that determines which rows to update, if a unique violation is found.
  * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testUpsertWithUniqueIndexConflict
  */
 fun <T : Table> T.upsert(
     vararg keys: Column<*>,
+    onUpdate: (UpsertBuilder.(UpdateStatement) -> Unit)? = null,
     onUpdateExclude: List<Column<*>>? = null,
     where: (SqlExpressionBuilder.() -> Op<Boolean>)? = null,
     body: T.(UpsertStatement<Long>) -> Unit
 ) = UpsertStatement<Long>(this, keys = keys, onUpdateExclude = onUpdateExclude, where = where?.let { SqlExpressionBuilder.it() }).apply {
+    onUpdate?.let { storeUpdateValues(it) }
     body(this)
     execute(TransactionManager.current())
 }
 
 @Deprecated(
-    "This `upsert()` with `onUpdate` parameter will be removed in future releases. " +
-        "Please use function `UpsertStatement.onUpdate()` in `body` lambda block instead.",
+    "This `upsert()` with `onUpdate` parameter that accepts a List will be removed in future releases. " +
+        "Please use `upsert()` with `onUpdate` parameter that takes an `UpdateStatement` lambda block instead.",
     level = DeprecationLevel.WARNING
 )
 fun <T : Table> T.upsert(
     vararg keys: Column<*>,
-    onUpdate: List<Pair<Column<*>, Expression<*>>>? = null,
+    onUpdate: List<Pair<Column<*>, Expression<*>>>,
```

Comment:
@e5l Making the `onUpdate` parameter of the old deprecated functions non-nullable means that any use of `upsert(onUpdate = listOf(...)) {}` shows a deprecation warning, while all other usages correctly resolve to the new function with the new parameter type.

I'm still good to drop all these deprecations and just break the parameter type if there's a reason not to do the change above.